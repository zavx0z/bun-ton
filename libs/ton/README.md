# Tests

```bash
bun test
```

[Bun test snapshots](https://bun.sh/guides/test/snapshot)
