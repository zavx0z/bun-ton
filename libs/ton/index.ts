export {
  readJettonMetadata,
  readJettonWalletMetadata,
  readNftItemMetadata,
  readNftMetadata,
  readContent,
} from "./src/metadata"
export { encodeOffChainContent } from "./src/encodeOffChain"
