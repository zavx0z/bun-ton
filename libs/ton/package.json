{
  "name": "@libs/ton",
  "private": true,
  "devDependencies": {
    "@aws-crypto/sha256-js": "^5.2.0",
    "@types/bun": "latest"
  },
  "dependencies":{},
  "peerDependencies": {
    "typescript": "^5.0.0"
  }
}