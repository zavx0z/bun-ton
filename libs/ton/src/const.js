/** OFF_CHAIN_CONTENT_PREFIX используется для обозначения off-chain контента в блокчейне TON.
 * Когда первый байт содержимого устанавливается в 0x01, это указывает, что данный контент
 * является ссылкой на off-chain ресурс, обычно на JSON файл, который содержит более
 * подробную информацию. Этот метод оптимизирует хранение данных в блокчейне, поскольку
 * в блокчейне хранится не полный контент, а лишь ссылка на внешний ресурс.
 */
export const OFFCHAIN_CONTENT_PREFIX = 0x01
/** ONCHAIN_CONTENT_PREFIX используется для обозначения on-chain контента в блокчейне TON.
 * Если первый байт содержимого устанавливается в 0x00, это указывает, что весь контент
 * хранится напрямую в блокчейне. Это применяется для контента, который должен быть постоянно
 * доступен и не требует внешних ссылок для его извлечения.
 */
export const ONCHAIN_CONTENT_PREFIX = 0x00
