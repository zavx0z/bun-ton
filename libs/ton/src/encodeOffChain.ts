import { beginCell, type Cell } from "@ton/core"

const bufferToChunks = (buff: Buffer, chunkSize: number) => {
  const chunks: Buffer[] = []
  while (buff.byteLength > 0) {
    chunks.push(buff.subarray(0, chunkSize))
    buff = buff.subarray(chunkSize)
  }
  return chunks
}
const makeSnakeCell = (data: Buffer): Cell => {
  const chunks = bufferToChunks(data, 127)
  if (chunks.length === 0) return beginCell().endCell()
  if (chunks.length === 1) return beginCell().storeBuffer(chunks[0]).endCell()
  let curCell = beginCell()
  for (let i = chunks.length - 1; i >= 0; i--) {
    const chunk = chunks[i]
    curCell.storeBuffer(chunk)
    if (i - 1 >= 0) {
      const nextCell = beginCell()
      nextCell.storeRef(curCell)
      curCell = nextCell
    }
  }
  return curCell.endCell()
}
export const encodeOffChainContent = (content: string) => {
  let data = Buffer.from(content)
  const offChainPrefix = Buffer.from([0x01])
  data = Buffer.concat([offChainPrefix, data])
  return makeSnakeCell(data)
}
