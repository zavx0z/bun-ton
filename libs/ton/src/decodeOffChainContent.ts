import { bitsToPaddedBuffer } from "@ton/core/dist/boc/utils/paddedBits"
import { Cell } from "@ton/ton"

const OFF_CHAIN_CONTENT_PREFIX = 0x01

export function flattenSnakeCell(cell: Cell) {
  let c: Cell | null = cell
  let res = Buffer.alloc(0)
  while (c) {
    let cs = c.beginParse()
    let data = bitsToPaddedBuffer(cs.loadBits(cs.remainingBits))
    res = Buffer.concat([res, data])
    c = c.refs[0]
  }
  return res
}
export function decodeOffChainContent(content: Cell) {
  let data = flattenSnakeCell(content)
  let prefix = data[0]
  if (prefix !== OFF_CHAIN_CONTENT_PREFIX) throw new Error(`Unknown content prefix: ${prefix.toString(16)}`)
  return data.slice(1).toString()
}
