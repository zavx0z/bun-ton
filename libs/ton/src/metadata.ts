import { Address, Slice, TonClient, TupleReader } from "@ton/ton"
import { Sha256 } from "@aws-crypto/sha256-js"
import { parseDict } from "@ton/core/dist/dict/parseDict"
import { bitsToPaddedBuffer } from "@ton/core/dist/boc/utils/paddedBits"
import type {
  Content,
  JettonMetaDataKeys,
  JettonMetadata,
  JettonOffchainMetadata,
  JettonOnChainMetadataSpec,
  JettonOnchainMetadata,
  NftMetadata,
} from "./types"

const jettonOnChainMetadataSpec: JettonOnChainMetadataSpec = {
  name: "utf8",
  description: "utf8",
  image: "ascii",
  decimals: "utf8",
  symbol: "utf8",
  image_data: undefined,
}

const ONCHAIN_CONTENT_PREFIX = 0x00
const OFFCHAIN_CONTENT_PREFIX = 0x01
const SNAKE_PREFIX = 0x00

const sha256 = (str: string) => {
  const sha = new Sha256()
  sha.update(str)
  return Buffer.from(sha.digestSync())
}

export const readContent = async (stack: TupleReader): Content => {
  const contentCell = stack.readCell()
  const contentSlice = contentCell.beginParse()
  switch (contentSlice.loadUint(8)) {
    case ONCHAIN_CONTENT_PREFIX:
      return {
        persistenceType: "onchain",
        ...parseJettonOnchainMetadata(contentSlice),
      }
    case OFFCHAIN_CONTENT_PREFIX:
      const { metadata, isIpfs } = await parseJettonOffchainMetadata(contentSlice)
      return {
        persistenceType: isIpfs ? "offchain_ipfs" : "offchain_private_domain",
        metadata,
      }
    default:
      throw new Error("Unexpected jetton metadata content prefix")
  }
}

export const readJettonWalletMetadata = async (client: TonClient, address: string) => {
  const walletData = await client.runMethod(Address.parse(address), "get_wallet_data")
  walletData.stack.skip(2)
  return readJettonMetadata(client, walletData.stack.readAddress().toString())
}

export const readNftItemMetadata = async (client: TonClient, address: string) => {
  const nftItemAddress = Address.parse(address)
  const nftData = await client.runMethod(nftItemAddress, "get_nft_data")
  nftData.stack.skip(1)
  const index = nftData.stack.readBigNumber()
  const collection_address = nftData.stack.readAddress()
  nftData.stack.skip(1)
  const individual_item_content = nftData.stack.readCell()
  const nftCollectionContent = await client.runMethod(collection_address, "get_nft_content", [
    { type: "int", value: index },
    { type: "cell", cell: individual_item_content },
  ])
  let linkSlice = nftCollectionContent.stack.readCell().beginParse()
  const pathBase = bitsToPaddedBuffer(linkSlice.loadBits(linkSlice.remainingBits)).toString()
  linkSlice = individual_item_content.beginParse()
  return fetch(
    linkSlice.remainingBits == 0
      ? pathBase
      : pathBase + "/" + bitsToPaddedBuffer(linkSlice.loadBits(linkSlice.remainingBits)).toString()
  ).then((data) => data.json())
}

export const readNftMetadata = async (client: TonClient, address: string): NftMetadata => {
  const nftCollectionAddress = Address.parse(address)
  const res = await client.runMethod(nftCollectionAddress, "get_collection_data")
  res.stack.skip(1)
  return await readContent(res.stack)
}

export const readJettonMetadata = async (client: TonClient, address: string): JettonMetadata => {
  const jettonMinterAddress = Address.parse(address)
  const res = await client.runMethod(jettonMinterAddress, "get_jetton_data")
  res.stack.skip(3)
  return await readContent(res.stack)
}

const parseJettonOnchainMetadata = (contentSlice: Slice): JettonOnchainMetadata => {
  const toKey = (str: string) => BigInt(`0x${str}`)
  const KEYLEN = 256
  let isJettonDeployerFaultyOnChainData = false
  const dict = parseDict(contentSlice.loadRef().beginParse(), KEYLEN, (s) => {
    let buffer = Buffer.from("")
    const sliceToVal = (s: Slice, v: Buffer, isFirst: boolean) => {
      s.asCell().beginParse()
      if (isFirst && s.loadUint(8) !== SNAKE_PREFIX) throw new Error("Only snake format is supported")
      const bits = s.remainingBits
      const bytes = bitsToPaddedBuffer(s.loadBits(bits))
      v = Buffer.concat([v, bytes])
      if (s.remainingRefs === 1) v = sliceToVal(s.loadRef().beginParse(), v, false)
      return v
    }
    if (s.remainingRefs === 0) {
      isJettonDeployerFaultyOnChainData = true
      return sliceToVal(s, buffer, true)
    }
    return sliceToVal(s.loadRef().beginParse(), buffer, true)
  })
  const res: { [s in JettonMetaDataKeys]?: string } = {}
  Object.keys(jettonOnChainMetadataSpec).forEach((k) => {
    const val = dict.get(toKey(sha256(k).toString("hex")))?.toString(jettonOnChainMetadataSpec[k as JettonMetaDataKeys])
    if (val) res[k as JettonMetaDataKeys] = val
  })
  return { metadata: res, isJettonDeployerFaultyOnChainData }
}

const parseJettonOffchainMetadata = async (contentSlice: Slice): JettonOffchainMetadata => {
  const bits = contentSlice.remainingBits
  const bytes = bitsToPaddedBuffer(contentSlice.loadBits(bits))
  const jsonURI = bytes.toString("ascii").replace("ipfs://", "https://ipfs.io/ipfs/")
  return { metadata: await (await fetch(jsonURI)).json(), isIpfs: /(^|\/)ipfs[.:]/.test(jsonURI) }
}
