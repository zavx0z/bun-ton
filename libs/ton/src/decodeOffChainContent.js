/**
 * Модуль для декодирования змеевидных ячеек с контентом.
 *
 * Змеевидная структура ячеек используется в TON для хранения
 * контента, который может быть большим и не помещаться в
 * одну ячейку. Внутри ячеек содержится последовательность
 * ячеек, каждая из которых может содержать часть контента.
 * Каждая из таких ячеек имеет общий префикс, который позволяет
 * определить, что это змеевидная ячейка.
 *
 * Этот модуль предоставляет функцию {@link decodeOffChainContent}, которая
 * принимает корневую ячейку такого контента и возвращает единый буфер,
 * содержащий все данные всех ячеек.
 *
 * @typedef {import("@ton/ton").Cell} Cell
 */

import { bitsToPaddedBuffer } from "@ton/core/dist/boc/utils/paddedBits"
import { OFFCHAIN_CONTENT_PREFIX } from "./const"

/** Преобразует змеевидную структуру ячеек в единый буфер.
 * @param {Cell} cell - Корневая ячейка змеевидной структуры.
 * @returns {Buffer} Буфер, содержащий данные всех ячеек.
 */
const flattenSnakeCell = (cell) => {
  let /** @type {Cell | null} */ c = cell // Начинаем с первой ячейки.
  let res = Buffer.alloc(0) // Инициализация результирующего буфера.
  while (c) {
    // Обходим все ячейки.
    let cs = c.beginParse() // Начинаем парсинг текущей ячейки.
    // Загружаем данные и преобразуем их в буфер.
    let data = bitsToPaddedBuffer(cs.loadBits(cs.remainingBits))
    // Конкатенируем данные текущей ячейки с предыдущими.
    res = Buffer.concat([res, data])
    c = c.refs[0] // Переходим к следующей ячейке в цепочке.
  }
  return res
}

/** Декодирует off-chain содержимое из ячейки.
 * @param {Cell} content - Корневая ячейка, содержащая off-chain данные.
 * @returns {string} Строковое представление данных, исключая префикс.
 */
export const decodeOffChainContent = (content) => {
  let data = flattenSnakeCell(content) // Формируем буфер из змеевидной структуры.
  let prefix = data[0] // Извлекаем префикс из данных. // Проверяем соответствие префикса ожидаемому значению.
  if (prefix !== OFFCHAIN_CONTENT_PREFIX) {
    // В случае несоответствия выбрасываем исключение.
    throw new Error(`Unknown content prefix: ${prefix.toString(16)}`)
  } // Возвращаем декодированное содержимое в виде строки, используя метод subarray.
  return Buffer.from(data).subarray(1).toString()
}
