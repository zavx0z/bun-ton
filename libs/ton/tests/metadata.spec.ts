import { describe, test, expect, beforeAll } from "bun:test"
import { readJettonMetadata, readJettonWalletMetadata, readNftItemMetadata, readNftMetadata } from "../src/metadata"
import { getHttpEndpoint } from "@orbs-network/ton-access"
import { TonClient } from "@ton/ton"

describe("metadata", () => {
  let endpoint
  let client: TonClient

  beforeAll(async () => {
    endpoint = await getHttpEndpoint({ network: "mainnet" })
    client = new TonClient({ endpoint })
  })

  test("parseJettonOffchainMetadata", async () => {
    const jettonData = await readJettonMetadata(client, "EQANasbzD5wdVx0qikebkchrH64zNgsB38oC9PVu7rG16qNB")
    expect(jettonData).toMatchSnapshot()
  })
  test("parseJettonWalletData", async () => {
    const jettonWalletData = await readJettonWalletMetadata(client, "EQBhxsx1cHE34hrAM-hRRv7c26G57pe2G6Iw1LTn_5hOuRoV")
    expect(jettonWalletData).toMatchSnapshot()
  })
  test("parseNftMetadata", async () => {
    const nftData = await readNftMetadata(client, "EQCvYf5W36a0zQrS_wc6PMKg6JnyTcFU56NPx1PrAW63qpvt")
    expect(nftData).toMatchSnapshot()
  })
  test("parseNftItemMetadata", async () => {
    const nftItemData = await readNftItemMetadata(client, "EQDTUKV5bMBh2SiL0eXgpO4XxPD1dp5DEpPR1fEIdHLJI40T")
    expect(nftItemData).toMatchSnapshot()
  })
})
