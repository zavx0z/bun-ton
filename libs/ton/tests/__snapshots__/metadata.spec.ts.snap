// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata parseJettonOffchainMetadata 1`] = `
{
  "isJettonDeployerFaultyOnChainData": false,
  "metadata": {
    "decimals": "8",
    "description": "Orbit Bridge Token on TON blockchain!",
    "image": "https://raw.githubusercontent.com/orbit-chain/bridge-token-image/main/ton/wbtc.png",
    "name": "Orbit Bridge Ton Wrapped BTC",
    "symbol": "oWBTC",
  },
  "persistenceType": "onchain",
}
`;

exports[`metadata parseJettonWalletData 1`] = `
{
  "isJettonDeployerFaultyOnChainData": false,
  "metadata": {
    "decimals": "9",
    "description": "This is first token of Alefman",
    "image": "https://alefmanvladimir.github.io/ton-study-counter-front/icon.png",
    "name": "Alefman Token",
    "symbol": "ALT",
  },
  "persistenceType": "onchain",
}
`;

exports[`metadata parseNftCollectionMetadata 1`] = `
{
  "metadata": {
    "cover_image": "https://richcats.app/api/collection/1/wide_cover",
    "description": "Rich Cats is the world's first customizable NFTs exclusively for Telegram.",
    "image": "https://richcats.app/api/collection/1/image",
    "name": "Rich Cats",
  },
  "persistenceType": "offchain_private_domain",
}
`;

exports[`metadata parseNftItemMetadata 1`] = `
{
  "attributes": [
    {
      "trait_type": "Type",
      "value": "Female",
    },
    {
      "trait_type": "Attributes",
      "value": "3",
    },
    {
      "trait_type": "Attribute 1",
      "value": "Green Earring",
    },
    {
      "trait_type": "Attribute 2",
      "value": "Blue Baseball Cap",
    },
    {
      "trait_type": "Attribute 3",
      "value": "Pink Eye Shadows",
    },
    {
      "trait_type": "Background",
      "value": "Radial 1",
    },
  ],
  "description": "Legendary Punks on the Legendary Blockchain",
  "image": "https://cloudflare-ipfs.com/ipfs/bafybeifav4uhztrmqmnao2vxtr6g2jbv62zhpmf3njry6mtti5khjtjncu/5.png",
  "name": "TON Punk #5",
}
`;
